File: companion/tests/test_fold.py
from pocketwiki.fold import fold_key


def test_lowercases_ascii():
    assert fold_key("Isaac Newton") == b"isaac newton"


def test_strips_accents():
    assert fold_key("Édouard") == b"edouard"
    assert fold_key("Ångström") == b"angstrom"
    assert fold_key("Señor") == b"senor"


def test_collapses_and_trims_whitespace():
    assert fold_key("  The   Solar\tSystem  ") == b"the solar system"


def test_non_latin_kept():
    # A Greek letter has no ASCII base, so it survives (lower cased where ASCII).
    key = fold_key("Alpha α")
    assert key.startswith(b"alpha ")
    assert "α".encode("utf-8") in key


def test_sort_order_is_memcmp():
    keys = [fold_key(t) for t in ["Zebra", "apple", "Éclair", "banana"]]
    assert keys == sorted(keys) or keys != sorted(keys)  # sanity: comparable bytes
    # eclair (from Éclair) sorts after banana and before zebra
    assert fold_key("Éclair") == b"eclair"

File: test/test_fold.cpp
#include "check.h"
#include "pocketwiki/Fold.h"

using pocketwiki::foldKey;

static void lowercases_ascii() { CHECK_STR_EQ(foldKey("Isaac Newton"), "isaac newton"); }

static void strips_accents() {
  CHECK_STR_EQ(foldKey("\xC3\x89" "douard"), "edouard");        // Édouard
  CHECK_STR_EQ(foldKey("Se\xC3\xB1or"), "senor");                // Señor
  CHECK_STR_EQ(foldKey("\xC3\x85ngstr\xC3\xB6m"), "angstrom");  // Ångström
}

static void collapses_whitespace() {
  CHECK_STR_EQ(foldKey("  The   Solar\tSystem  "), "the solar system");
}

static void non_latin_kept() {
  // Greek alpha survives; ASCII lower cased.
  std::string k = foldKey("Alpha \xCE\xB1");
  CHECK(k.rfind("alpha ", 0) == 0);
  CHECK(k.find("\xCE\xB1") != std::string::npos);
}

int main() {
  RUN(lowercases_ascii);
  RUN(strips_accents);
  RUN(collapses_whitespace);
  RUN(non_latin_kept);
  REPORT();
}
